cpp/src/arrow/dataset/file_json_test.cc

```diff
@@ -0,0 +1,216 @@
+      RETURN_NOT_OK(WriteJson(*scalar, &sink));
+      ss << "\n";
+    }
+    return Buffer::FromString(ss.str());
+  }
+
+  static std::shared_ptr<FormatType> MakeFormat() {
+    return std::make_shared<FormatType>();
+  }
+
+ private:
+  static Result<std::vector<std::shared_ptr<StructScalar>>> ToScalars(
+      RecordBatchReader* reader) {
+    std::vector<std::shared_ptr<StructScalar>> scalars;
+    for (auto maybe_batch : *reader) {
+      ARROW_ASSIGN_OR_RAISE(auto batch, maybe_batch);
+      ARROW_ASSIGN_OR_RAISE(auto array, batch->ToStructArray());
+      for (int i = 0; i < array->length(); ++i) {
+        ARROW_ASSIGN_OR_RAISE(auto scalar, array->GetScalar(i));
+        scalars.push_back(checked_pointer_cast<StructScalar>(std::move(scalar)));
+      }
+    }
+    return scalars;
+  }
+};
+
+class TestJsonFormat
+    : public FileFormatFixtureMixin<JsonFormatHelper, json::kMaxParserNumRows> {};
+
+class TestJsonFormatScan : public FileFormatScanMixin<JsonFormatHelper> {};
+
+class TestJsonFormatScanNode : public FileFormatScanNodeMixin<JsonFormatHelper> {
+  void SetUp() override { internal::Initialize(); }
+
+  const FragmentScanOptions* GetFormatOptions() override { return &json_options_; }
+
+ protected:
+  JsonFragmentScanOptions json_options_;
+};
+
+TEST_F(TestJsonFormat, Equals) {
+  JsonFileFormat format;
+  ASSERT_TRUE(format.Equals(JsonFileFormat()));
+  ASSERT_FALSE(format.Equals(DummyFileFormat()));
+}
+
+TEST_F(TestJsonFormat, IsSupported) { TestIsSupported(); }
+TEST_F(TestJsonFormat, Inspect) { TestInspect(); }
+TEST_F(TestJsonFormat, FragmentEquals) { TestFragmentEquals(); }
+TEST_F(TestJsonFormat, InspectFailureWithRelevantError) {
+  TestInspectFailureWithRelevantError(StatusCode::Invalid, "JSON");
+}
+TEST_F(TestJsonFormat, CountRows) { TestCountRows(); }
+
+TEST_P(TestJsonFormatScan, Scan) { TestScan(); }
+TEST_P(TestJsonFormatScan, ScanBatchSize) { TestScanBatchSize(); }
+TEST_P(TestJsonFormatScan, ScanProjected) { TestScanProjected(); }
+TEST_P(TestJsonFormatScan, ScanWithDuplicateColumnError) {
+  TestScanWithDuplicateColumnError();
+}
+TEST_P(TestJsonFormatScan, ScanWithVirtualColumn) { TestScanWithVirtualColumn(); }
+TEST_P(TestJsonFormatScan, ScanWithPushdownNulls) { TestScanWithPushdownNulls(); }
+TEST_P(TestJsonFormatScan, ScanProjectedMissingCols) { TestScanProjectedMissingCols(); }
+TEST_P(TestJsonFormatScan, ScanProjectedNested) { TestScanProjectedNested(); }
+
+INSTANTIATE_TEST_SUITE_P(TestJsonScan, TestJsonFormatScan,
+                         ::testing::ValuesIn(TestFormatParams::Values()),
+                         TestFormatParams::ToTestNameString);
+
+TEST_P(TestJsonFormatScanNode, Scan) { TestScan(); }
+TEST_P(TestJsonFormatScanNode, ScanMissingFilterField) { TestScanMissingFilterField(); }
+TEST_P(TestJsonFormatScanNode, ScanProjected) { TestScanProjected(); }
+TEST_P(TestJsonFormatScanNode, ScanProjectedMissingColumns) {
+  TestScanProjectedMissingCols();
+}
+
```

Comment:
Just scanning through the CSV tests I wonder if there are a few JSON-specific tests that might be nice to have:

 * Confirm that alternate read/parse options (e.g. alternate block size or newlines allowed) get passed in correctly
 * Does the JSON reader handle the the presence of a unicode BOM well?  Maybe not as big an issue here where we have to explicitly handle it for the CSV reader.

Reply:
The reader doesn't explicitly handle BOM, but I suspect that rapidjson does, as it's at least [referenced in their docs](https://rapidjson.org/classrapidjson_1_1_encoding.html). Either way, I added a test for it (which currently passes).